### ISSUE TITLE
Pondering makes no advantage even if LZ plays with itself.

### ISSUE THREAD

**Issue by bjiyxo** [gcp/leela-zero#918]
As title. 
The test was on 7900X and 2*1080Ti. Each player used 1 1080Ti, and one used ponder(as default) and the other used noponder. I tested 2 games(Both of them play black or white once.), and surprisingly, the one with ponder doesn't have any advantage and lose both of games even if LZ is playing with itself. 
Is pondering broken?

**Comment by billyswong** [comment]
No conclusions can be made by just 2 games. Even if pondering win 60%, there is still 16% chance for the first 2 games to lost, around the probability of throwing a dice and get 1 point.

**Comment by bjiyxo** [comment]
I agree with your stamement. But I've observed 2 games and the win rate from both side is around 40-60%. Conversely, I tested 2 gpus vs 1 gpu and I can see the win rate is slightly climbing from 2 gpus' side. Well, I can test more games if needed. Or you may suggest some advice that I need to pay attention to.

**Comment by Cabu** [comment]
Pondering, like I understand it barely make leela stronger. The advantage for me is that when leela play against an opponent that take a lot of time to answer a move, it get some time to prepare some possible answers to possible opponent moves making lella play instantaneously or more quickly when the opponent finnaly play its move. And by that, take less time on its clock during timed games. 
I have not checked the code, but I think that when pondering, leela is not limited by the number of playouts/visits and can look deeper in the tree to find a better move. But the probability of the opponent playing the next best move (execpt if the opponent is an engine or high dans player) make that search go to waste most of the time.

**Comment by tterava** [comment]
Currently the way pondering is set up, it basically just speeds up the games. If the opponent makes a move LZ is expecting, it will instantly answer because it's already done the necessary amount of playouts. You only get an advantage from pondering if opponent uses more time during his move than LZ would AND opponent makes the move LZ is expecting.

**Comment by wctgit** [comment]
I wonder if, during pondering mode, the search could be set to be more exploratory, to allow it to branch out a bit more and anticipate other possible main lines than just the top one or two. This might give pondering a greater advantage, possibly.

**Comment by WhiteHalmos** [comment]
The problem is that all those extra lines are discarded when the opponent chooses just one, and perhaps not even one of those lines. So the vast majority of pondering does nothing to benefit the subsequent search.

**Comment by Cabu** [comment]
Pondering is a bet even if it is wrong 90% of the time, is still a better usage of the CPU than letting it rest doing nothing.

**Comment by Dorus** [comment]
How much time did you give the bots? I hope you did not limit it to playouts or visits?

**Comment by sethtroisi** [comment]
DATA

supervised weights, ~1 seconds per move
```
ponder v noponder (32/200 games)
board size: 19   komi: 7.5
           wins              black         white       avg cpu
ponder       18 56.25%       7  43.75%     11 68.75%    179.87
noponder     14 43.75%       5  31.25%     9  56.25%     90.73
                             12 37.50%     20 62.50%
```

**Comment by bjiyxo** [comment]
Thanks for the test. I used 20s/move and my n/s is around 3000~5000(1*1080Ti). I'll test 100 games under same condition now. Maybe tomorrow I will give you results.

**Comment by bjiyxo** [comment]
DATA
weight files:92dd0397, 20 s per move(speed is ~3000~5000n/s)
`ponder vs noponder (20 games)`
`                  wins           black          white`
`ponder       8                3                5`
`noponder   12              5                7`
`                                    8                12`
So it seems pondering doesn't make Leelaz stronger.
I will do the same test with 53f4cc79 tomorrow.

For those who are interested, the following is the 20 sgf files.
[PonderingTest.zip](https://github.com/gcp/leela-zero/files/1764090/PonderingTest.zip)
1Gnp means 1 GPU with no pondering and 1G means 1 GPU with pondering as default.
1G:-g -t3 -w 92dd0397bb12dad2e1a721230ea89a5f24967e476eca440952c64733cd9434b7 --gpu 0
time_settings 0 20 1
1Gnp:-g -t3 -w 92dd0397bb12dad2e1a721230ea89a5f24967e476eca440952c64733cd9434b7 --gpu 1 --noponder
time_settings 0 20 1

**Comment by Dorus** [comment]
8-12 is nowhere near statistical significant.

**Comment by bjiyxo** [comment]
@Dorus Yes. But I need to test ~20s/move because the real games are just like this, so the test will be long. Well, I will test around 5 weight files and each one will be tested about 20 games. So we will see the results.
PS: there is no -time for leelaz and I need to test every game by my hand and use time_settings. Can anybody add -time?

**Comment by wctgit** [comment]
Isn't that handled using GTP commands?

**Comment by bjiyxo** [comment]
@wctgit I'm not sure what you mean. I used to test AI's selfplay by gogui-twogtp for multiple games. However, I don't know how to use time_settings on twogtp. So I have to start every game by my hand.

**Comment by roy7** [comment]
@bjiyxo https://www.mankier.com/1/gogui-twogtp says it has a command line option --time and it will pass that info to the bot using time_settings for you. Looks like --time 5m+30s is what you'd do for a human style time control?

**Comment by bjiyxo** [comment]
@roy7 --time works only if the go program support it. Do you try it on lz by yourself?

**Comment by roy7** [comment]
--time is a command you give to gogui-twogtp, not leela zero. It will communicate time controls via time_settings to leela zero automatically.

**Comment by roy7** [comment]
Try -verbose -time 5m+30s/1 which will print all gtp traffic back and forth. You should see the appropriate time_settings command going to leela zero.

**Comment by bjiyxo** [comment]
@roy7 Thanks a lot. It works. I'm gonna finished the experiment through twogtp.

**Comment by marcocalignano** [comment]
Or maybe just use `validation` so you have SPRT to backup your statments.

**Comment by bjiyxo** [comment]
DATA
weight files:53f4cc79, 20 s per move(speed is ~ 3000 ~ 5000n/s)

```
ponder vs noponder (20 games)
		wins		black		white
ponder		11		4		7
noponder	9		3		6
				7               13
```

For those who are interested, the followings are the 20 sgf files.
[PonderingTest2.zip](https://github.com/gcp/leela-zero/files/1768021/PonderingTest2.zip)

1Gnp means 1 GPU with no pondering and 1G means 1 GPU with pondering as default.
1G:-g -t3 -w 53f4cc79987649d1126ff55aa4c63f6013f29d13164c247b669e676ce8c4838c --gpu 0
1Gnp:-g -t3 -w 53f4cc79987649d1126ff55aa4c63f6013f29d13164c247b669e676ce8c4838c --gpu 1 --noponder
4 games are tested by sabaki, and the other 16 games are tested by gogui-twogtp(-time 1s+20s/1).

So the total games(including the last time)
```
ponder vs noponder (20 games)
		wins		black		white
ponder		19		7		12
noponder	21		8		13
				15               25
```

**Comment by wctgit** [comment]
Even 40 games is too few to draw any confident conclusions. Try for 400 to get what this project's community considers almost a 'gold standard'. If that will take a long time, you can at least start the process and post updates as more results roll in.

Also, pondering would likely be most noticeably useful during games where the opponent has a significant amount of time, so using 1 second per turn seems way too low to me. I would think something more like at least 10s per move.

Also, you might need to be careful about comparing pondering vs. no pondering on the same machine, even with multiple GPUs. During the 'pondering' bot's turns, all computation resources are available to the engine, with no contention from anyone else. But during the 'no pondering' bot's turns, even if you assign each bot to a different GPU, there will still be some contention over computational resources, such as the CPU (even with multiple CPUs or multiple cores), thread/process context switching, thread communication, etc. As far as I'm aware, there's no perfect separation between parallel processes on the same machine. This might make the 'no pondering' bot seem artificially weaker, just because it's only getting 80% of a GPU or whatever.

You might want to think about doing this test using two separate, similar machines, or better yet, two similar machines for the two bots, and a third machine to act as the game/tournament server. I think GTP is designed with this capability in mind, so it shouldn't be too much of a problem, AFAIK, assuming you have access to at least two similar machines.

**Comment by killerducky** [comment]
@bjiyxo can you add something like `-l ponder.log` and `-l noponder.log` to the LZ command line, run one game, and post the log files?

**Comment by bjiyxo** [comment]
@wctgit I did it in 20s byoyomi. I have only a machine with 2x1080Ti, and if I separate a 1080Ti to another machine, I have to match the cpu simultaneously. It's impossible for me to do this.
My plan is doing 60 games to the next weights and use the same method, and I'll add -l as @killerducky mention.
